### PR TITLE
Issue #20 : GetContextFromConstantExpression now handles embedded contexts

### PR DESCRIPTION
**Pull request by jcchalte**
This is a correction for the bug #20 .

As mentioned in the bug, the existing code assume that the left part of the method call is a MemberAccess on a constant. It was the case in VS2013 but it is not the case anymore, as since VS2015, the context can be embedded in a field of the constant.

The corrected code does not assume anymore anything but tries to retrieve the left part value recursively until it finds the constant.

### CODE REVIEW

File: src/EntityFramework/Internal/Linq/DbQueryVisitor.cs [hunk 1]
```diff
@@ -120,15 +120,60 @@ private static DbContext GetContextFromConstantExpression(Expression expression,
...
+        // If the expression is a constant, it returns the constant value.
+        // If the expression is a field or property access on an expression, it returns the value of it recursively.
+        // Otherwise it returns null
+        // </summary>
+        // <param name="expression">The expression</param>
+        // <returns> The expression value. </returns>
+        private static object GetExpressionValue(Expression expression)
+        {
+            //If the given expression is a constant, we just return its value
+            ConstantExpression constantExpression = expression as ConstantExpression;
```
Comment (maumar) [inline src/EntityFramework/Internal/Linq/DbQueryVisitor.cs]: use 'var' here

File: src/EntityFramework/Internal/Linq/DbQueryVisitor.cs [hunk 2]
```diff
@@ -120,15 +120,60 @@ private static DbContext GetContextFromConstantExpression(Expression expression,
...
+            if (constantExpression != null)
+            {
+                return constantExpression.Value;
+            }
+
+            //If the given expression is a member access on an inner expression, we recursively retrieve the value of the inner expression, and get the member value from it.
+            MemberExpression memberExpression = expression as MemberExpression;
```
Comment (maumar) [inline src/EntityFramework/Internal/Linq/DbQueryVisitor.cs]: use 'var' here

File: src/EntityFramework/Internal/Linq/DbQueryVisitor.cs [hunk 3]
```diff
@@ -120,15 +120,60 @@ private static DbContext GetContextFromConstantExpression(Expression expression,
...
+            if (memberExpression != null)
+            {
+                var asField = memberExpression.Member as FieldInfo;
+                if (asField != null)
                 {
-                    return GetContextFromMember(member, value);
+                    object innerValue = GetExpressionValue(memberExpression.Expression);
```
Comment (maumar) [inline src/EntityFramework/Internal/Linq/DbQueryVisitor.cs]: use 'var' here

File: src/EntityFramework/Internal/Linq/DbQueryVisitor.cs [hunk 4]
```diff
@@ -120,15 +120,60 @@ private static DbContext GetContextFromConstantExpression(Expression expression,
                 return GetContextFromMember(member, null);
             }
 
-            var constant = expression as ConstantExpression;
-            if (constant != null)
+            //Retrieve the context value from the encapsulated scope
+            object value = GetExpressionValue(expression);
```
Comment (maumar) [inline src/EntityFramework/Internal/Linq/DbQueryVisitor.cs]: use 'var' here

**Comment by jcchalte** [comment]
OK, thank you for your review. I will apply those changes asap and come back to you.

**Comment by jcchalte** [comment]
It should be OK now.

**Comment by maumar** [comment]
@jcchalte I have merged your pull request. Thank you for the contribution! Relevant commits: https://github.com/aspnet/EntityFramework6/commit/926744d447e3ddfa78a7f4af6c095373b7c8213a https://github.com/aspnet/EntityFramework6/commit/6463c603fe8b9f8132c7766c446fc76f58bf748e https://github.com/aspnet/EntityFramework6/commit/e7628ba43584da9d406d00726ecc6a96846f1209

**Pull request closed without merging**